File: sb_editor/sb_editor/Custom Controls/MultiSelListBox.vb
Public Class MultiSelListBox
    Inherits ListBox

    Private MouseDownOnIndex As Integer
    Private bMouseDownOnSelection As Boolean
    Private bMouseDownOutsideSelection As Boolean
    Private Const WM_LBUTTONDOWN As Integer = &H201
    Private Const WM_LBUTTONUP As Integer = &H202
    Private Const WM_MOUSEMOVE As Integer = &H200
    Private Const MK_LBUTTON As Integer = &H1&
    Private lastClickedPosition As Point
    Public Property DragDropEffectVal As DragDropEffects

    Protected Overrides Sub OnCreateControl()
        lastClickedPosition = Point.Empty
        MyBase.OnCreateControl()
    End Sub

    Protected Overrides Sub WndProc(ByRef m As System.Windows.Forms.Message)
        Select Case m.Msg
            Case WM_LBUTTONDOWN
                Dim pt As New Point(m.LParam.ToInt32)
                MouseDownOnIndex = Me.IndexFromPoint(pt)
                If Me.SelectedItems.Count >= 1 _
                    And Me.SelectedIndices.Contains(MouseDownOnIndex) _
                        And m.WParam = MK_LBUTTON Then
                    bMouseDownOnSelection = True
                    lastClickedPosition = pt
                    Return
                Else
                    bMouseDownOutsideSelection = True
                    lastClickedPosition = Point.Empty
                    MyBase.WndProc(m)
                End If
            Case WM_MOUSEMOVE
                If bMouseDownOnSelection Then
                    Dim pt As New Point(m.LParam.ToInt32)
                    If Math.Abs(pt.X - lastClickedPosition.X) < 3 OrElse
                        Math.Abs(pt.Y - lastClickedPosition.Y) < 3 Then
                        Me.DoDragDrop(Me.SelectedItems, DragDropEffectVal)
                    End If
                End If
                bMouseDownOnSelection = False
                MyBase.WndProc(m)
            Case WM_LBUTTONUP
                Dim pt As New Point(m.LParam.ToInt32)
                If MouseDownOnIndex = Me.IndexFromPoint(pt) _
                    And m.WParam = 0 And Not bMouseDownOutsideSelection Then
                    Dim down As New Message With {
                        .HWnd = m.HWnd,
                        .Msg = WM_LBUTTONDOWN,
                        .WParam = m.WParam,
                        .LParam = m.LParam,
                        .Result = IntPtr.Zero
                    }
                    MyBase.WndProc(down)
                    lastClickedPosition = Point.Empty
                End If
                bMouseDownOutsideSelection = False
                MyBase.WndProc(m)
            Case Else
                MyBase.WndProc(m)
        End Select
    End Sub

End Class

File: Solutions/EuroSound Editor/Custom_Controls/MultiSelListBox.cs
using System;
using System.Drawing;
using System.Windows.Forms;

public partial class MultiSelListBox : ListBox
{
    private int MouseDownOnIndex;
    private bool bMouseDownOnSelection;
    private bool bMouseDownOutsideSelection;
    private const int WM_LBUTTONDOWN = 0x201;
    private const int WM_LBUTTONUP = 0x202;
    private const int WM_MOUSEMOVE = 0x200;
    private const int MK_LBUTTON = 0x1;
    private Point lastClickedPosition;
    public DragDropEffects DragDropEffectVal { get; set; }

    protected override void OnCreateControl()
    {
        lastClickedPosition = Point.Empty;
        base.OnCreateControl();
    }

    protected override void WndProc(ref Message m)
    {
        switch (m.Msg)
        {
            case WM_LBUTTONDOWN:
                Point pt = new Point(m.LParam.ToInt32());
                MouseDownOnIndex = IndexFromPoint(pt);
                if (SelectedItems.Count >= 1 & SelectedIndices.Contains(MouseDownOnIndex) & (int)m.WParam == MK_LBUTTON)
                {
                    bMouseDownOnSelection = true;
                    lastClickedPosition = pt;
                    return;
                }
                else
                {
                    bMouseDownOutsideSelection = true;
                    lastClickedPosition = Point.Empty;
                    base.WndProc(ref m);
                }

                break;
            case WM_MOUSEMOVE:
                if (bMouseDownOnSelection)
                {
                    pt = new Point(m.LParam.ToInt32());
                    if (Math.Abs(pt.X - lastClickedPosition.X) < 3 || Math.Abs(pt.Y - lastClickedPosition.Y) < 3)
                    {
                        DoDragDrop(SelectedItems, DragDropEffectVal);
                    }
                }
                bMouseDownOnSelection = false;
                base.WndProc(ref m);
                break;
            case WM_LBUTTONUP:
                pt = new Point(m.LParam.ToInt32());
                if (MouseDownOnIndex == IndexFromPoint(pt) & (int)m.WParam == 0 & !bMouseDownOutsideSelection)
                {
                    Message down = new Message()
                    {
                        HWnd = m.HWnd,
                        Msg = WM_LBUTTONDOWN,
                        WParam = m.WParam,
                        LParam = m.LParam,
                        Result = IntPtr.Zero
                    };
                    base.WndProc(ref down);
                    lastClickedPosition = Point.Empty;
                }
                bMouseDownOutsideSelection = false;
                base.WndProc(ref m);
                break;
            default:
                base.WndProc(ref m);
                break;
        }
    }

}